qiskit/transpiler/coupling.py

```diff
@@ -412,6 +412,19 @@ def __str__(self):
             string += "]"
         return string
 
+    def __eq__(self, other):
+        """Check if the graph in other is isomorphic to the graph in self.
+
+        Args:
+            other (CouplingMap): The other coupling map.
+
+        Returns:
+            bool: Whether or not other is isomorphic to self.
+        """
+        if not isinstance(other, CouplingMap):
+            return False
+        return rx.is_isomorphic(other.graph, self.graph)
```

Comment:
Something I was thinking about overnight is whether this is actually the behavior we want on `__eq__` or not. An isomorphic graph is structurally equivalent, but that might not be what people expect with equals. I know For example, the case I was thinking of was if you had two coupling maps:

```python
cmap_a = CouplingMap([[0, 1], [0, 2], [2, 3], [1, 3], [4, 5], [5, 6], [6, 7], [7, 0]])
cmap_b = CouplingMap([[0, 1], [1, 2], [2, 3], [3, 4], [4, 5], [4, 6], [6, 7], [5, 7]])
```

with this PR `cmap_a == cmap_b` would be `True`, but would the normal expectation be for that to be `False`?
I'm wondering if changing this to be a comparison of the edge lists instead of isomorphism would be a better `__eq__` definition. If so we can add an `equiv()` or `isomorphic()` method to capture this use case too.
I know that I was the one who suggested this approach for `__eq__` originally, but do you have any thoughts here?

Reply:
I agree with this - I think the expectation is that `__eq__` should be quite strict.  Especially because the `CouplingMap` implicitly relates _physical_ qubits to their connections - if I have `cm1 == cm2` and `cm1` has an edge between physical qubits 0 and 1, I'd expect to be able to interact physical qubits 0 and 1 on the machine backed by `cm2` too, but if we relax `__eq__` to isomorphism, that won't be the case.

imo, especially from my pain dealing with other places in Qiskit where this isn't the case, `__eq__` should be a fairly strict but cheap method wherever possible.

Reply:
I agree with the point that node labelling matters here. @mtreinish you're suggesting edge list comparison because it uses the actual labels? I guess simplest thing would be to convert edge lists to sets and check equality of sets. Are there any ordering guarantees on the edge lists, or is there no way around something like the set comparison?

Reply:
Ah that's a good point, in it's current form an edge list comparison is fine because the graph needs to be connected. But after #9710 this won't be the case and there can be disconnected nodes in the graph. But I assume this will merge before #9710 so lets assume we don't have to worry about disconnected nodes for this PR and I'll update `__eq__` in #9710 after this merges.

As for the edge list order I  think you're correct that wrapping it in a `set` is probably more robust to ignore edge  order. The returned edges in `PyDiGraph.edge_list()` are always in insertion/creation order, but that could differ between graphs that I think should should be `==`. So we should do the comparison in an order agnostic way.